### PR TITLE
mbedtls_ssl_get_alert(): getter for fatal alerts

### PR DESCRIPTION
**Pull request by ng-gsmk**
## Description

Even though the TLS RFCs do not mandate libraries to expose *Error Alerts* (as defined in RFC8446 6.2 for TLS 1.3 and in RFC5246 7.2.2 for TLS 1.2) to the user, there are use cases when it is handy to get the actual last received fatal error instead of a generic one. For instance this enables the user to differ between received fatal errors in case `mbedtls_ssl_handshake()`, `mbedtls_ssl_handshake_step()` or `mbedtls_ssl_read()` returned `MBEDTLS_ERR_SSL_FATAL_ALERT_MESSAGE`.

This changesets stores the last incoming fatal alert in `mbedtls_ssl_context` and provides `mbedtls_ssl_get_alert()` as a getter for retrieving it. Another option would be to provide a callback mechanisms for all kinds of alerts (not only fatals) but for simplicity I discarded this option.

## PR checklist

- [x] **changelog** provided
- [x] **development PR** provided
- [x] **TF-PSA-Crypto PR** not required
- [x] **framework PR** not required
- [x] **3.6 PR** not required because of new feature?
- **tests**  provided

### CODE REVIEW

**Comment by bjwtaylor** [comment]
Hi @ng-gsmk, thanks for uploading this. There are currently CI failures, are you able to look into them?

**Comment by ng-gsmk** [comment]
I removed `mbedtls_ssl_conf_rng()` from the test case as it seems not needed and not available on all target platforms but I was wondering about the following report:

```
<problems_with_types severity="Medium">
  <header name="ssl.h">
    <type name="struct mbedtls_ssl_context">
      <problem id="Added_Field_And_Layout">
        <change target="in_alert_recv">Field @target has been added at the middle position of this structural type.</change>
        <effect>Layout of structure fields has been changed and therefore fields at higher positions of the structure definition may be incorrectly accessed by applications.</effect>
      </problem>
      <problem id="Added_Field_And_Layout">
        <change target="in_alert_type">Field @target has been added at the middle position of this structural type.</change>
        <effect>Layout of structure fields has been changed and therefore fields at higher positions of the structure definition may be incorrectly accessed by applications.</effect>
      </problem>
      </type>
  </header>
</problems_with_types>

<problems_with_types severity="Low">
  <header name="ssl.h">
    <type name="struct mbedtls_ssl_context">
      <problem id="Added_Field">
        <change target="in_alert_recv">Field @target has been added to this type.</change>
        <effect>This field will not be initialized or used by old client applications.</effect>
      </problem>
      <problem id="Added_Field">
        <change target="in_alert_type">Field @target has been added to this type.</change>
        <effect>This field will not be initialized or used by old client applications.</effect>
      </problem>
      </type>
  </header>
</problems_with_types>
```

What are the best practices in mbedtls when adding a new fields to mbedtls_ssl_context. I grouped it with the other alert stuff but this report indicates I should add it to the end?

**Comment by gilles-peskine-arm** [comment]
It's ok to change structures in the development branch and there's no particular reason to add new fields at the end. In Mbed TLS, struct types that are public are usually caller-allocated, so adding a field is an ABI change, but fields are usually private, and moving a private field isn't an ABI change. The tool we use for ABI comparison is not aware of how we use structures and which fields we consider to be private.

We tend to use the following considerations to order fields:

* Fields that are accessed often go earlier. If there's a field that's accessed much more often than others, it goes first. That tends to result in smaller code.
* Try to avoid gaps due to alignment requirements. An easy way is to put fields in decreasing order of alignment.
* But don't follow the guidelines above blindly. Try to group related fields, to make things easier on maintainers.

**Comment by ng-gsmk** [comment]
@bjwtaylor: PR tests seem to pass. The interface stability tests fail but this is expected I would say.

File: library/ssl_tls.c [hunk 1]
```diff
@@ -1297,6 +1297,8 @@ void mbedtls_ssl_session_reset_msg_layer(mbedtls_ssl_context *ssl,
         memset(ssl->in_buf, 0, in_buf_len);
     }
 
+    ssl->in_alert_recv = 0;
+
```
Comment (minosgalanakis) [inline library/ssl_tls.c]: Extra whitespace here and we are missing clearing the `ssl->in_alert_type`

File: library/ssl_msg.c [hunk 2]
```diff
@@ -5016,6 +5018,14 @@ int mbedtls_ssl_send_alert_message(mbedtls_ssl_context *ssl,
     return 0;
 }
 
+int mbedtls_ssl_get_alert(mbedtls_ssl_context *ssl)
+{
+    if (ssl == NULL || ssl->in_alert_recv != 1) {
+        return MBEDTLS_ERR_SSL_BAD_INPUT_DATA;
```
Comment (minosgalanakis) [inline library/ssl_msg.c:5023]: Not sure how I feel about it. The change is small to warrant a new discrete allert type for when there is no allert, but on the other side, a valid ssl context containing no allert is technically not a bad input, whereas it is a bad input for `mbedtls_ssl_get_alert`. 

An option would be to return `MBEDTLS_SSL_ALERT_MSG_CLOSE_NOTIFY` when `ssl->in_alert_recv != 1` but I also does not feel right error as well. 

@ronald-cron-arm wdyt?

File: tests/suites/test_suite_ssl.function [hunk 3]
```diff
@@ -5936,3 +5936,45 @@ exit:
...
+    ssl.in_msg[0] = MBEDTLS_SSL_ALERT_LEVEL_FATAL;
+    ssl.in_msg[1] = MBEDTLS_SSL_ALERT_MSG_HANDSHAKE_FAILURE;
+
+    /* import prepared fatal alert and test getter */
+    TEST_ASSERT(mbedtls_ssl_handle_message_type(&ssl) == MBEDTLS_ERR_SSL_FATAL_ALERT_MESSAGE);
+    TEST_ASSERT(mbedtls_ssl_get_alert(&ssl) == MBEDTLS_SSL_ALERT_MSG_HANDSHAKE_FAILURE);
+
+    /* Reset the session and check that no alert is present*/
+    mbedtls_ssl_session_reset_msg_layer(&ssl, 0);
+    TEST_ASSERT(mbedtls_ssl_get_alert(&ssl) == MBEDTLS_ERR_SSL_BAD_INPUT_DATA);
```
Comment (minosgalanakis) [inline tests/suites/test_suite_ssl.function]: Need to test the case of `TEST_ASSERT(mbedtls_ssl_get_alert(NULL) == MBEDTLS_ERR_SSL_BAD_INPUT_DATA);`

File: library/ssl_tls.c (hunk 1 above)
Comment (ng-gsmk) [reply library/ssl_tls.c]: I'm sorry but I can't follow what you mean with the extra whitespace? Is there an extra whitespace in the changeset missing or one that should be removed?

Regarding the extra clearing: I followed the way how send alert was implemented (there is no clearing for the type for a reset neither) but added it now with 33bd8f8.

File: tests/suites/test_suite_ssl.function (hunk 3 above)
Comment (ng-gsmk) [reply tests/suites/test_suite_ssl.function]: Added an assert for `mbedtls_ssl_get_alert(NULL)` with 6140cfb

**Comment by ng-gsmk** [comment]
@minosgalanakis thanks for the review. As requested I rebased the PR on top of development and fixed some remarks (see in-place comments).

File: library/ssl_tls.c (hunk 1 above)
Comment (minosgalanakis) [reply library/ssl_tls.c]: With regards to the whitespace I meant that instead of doing 

    ssl->in_alert_recv = 0;
    ssl->in_alert_type = 0;

    ssl->send_alert = 0;
    
We could have
    
    ssl->in_alert_recv = 0;
    ssl->in_alert_type = 0;
    ssl->send_alert = 0;
    
But do not push a change just for that, wait for the second reviewer's comments.

Also same applies with rebase, I would recommend to do it after both reviewers have approved a pr, so you do not have to redo it every time the base branch has created a conflict ;)

Thanks for getting on it so quickly.

File: library/ssl_msg.c [hunk 4]
```diff
@@ -5015,6 +5017,14 @@ int mbedtls_ssl_send_alert_message(mbedtls_ssl_context *ssl,
     return 0;
 }
 
+int mbedtls_ssl_get_alert(mbedtls_ssl_context *ssl)
```
Comment (ronald-cron-arm) [inline library/ssl_msg.c]:
```suggestion
int mbedtls_ssl_get_alert(const mbedtls_ssl_context *ssl)
```

File: include/mbedtls/ssl.h [hunk 5]
```diff
@@ -4911,6 +4918,22 @@ int mbedtls_ssl_write(mbedtls_ssl_context *ssl, const unsigned char *buf, size_t
...
+ * \return         The alert description type (MBEDTLS_SSL_ALERT_MSG_*) if a fatal
+ *                 alert has been received or MBEDTLS_ERR_SSL_BAD_INPUT_DATA
+ *
+ * \note           This function can be used in case mbedtls_ssl_handshake(),
+ *                 mbedtls_ssl_handshake_step() or mbedtls_ssl_read() returned
+ *                 MBEDTLS_ERR_SSL_FATAL_ALERT_MESSAGE to get the actual alert
+ *                 description type.
+ */
+int mbedtls_ssl_get_alert(mbedtls_ssl_context *ssl);
+
```
Comment (ronald-cron-arm) [inline include/mbedtls/ssl.h:4936]:
```suggestion
/**
 * \brief           Get the last received fatal alert
 *
 * \param ssl       SSL context
 *
 * \return         The alert description type (MBEDTLS_SSL_ALERT_MSG_*) if a fatal
 *                 alert has been received, MBEDTLS_ERR_SSL_BAD_INPUT_DATA otherwise.
 *
 * \note           This function can be used in case mbedtls_ssl_handshake(),
 *                 mbedtls_ssl_handshake_step() or mbedtls_ssl_read() returned
 *                 MBEDTLS_ERR_SSL_FATAL_ALERT_MESSAGE to get the actual alert
 *                 description type.
 */
int mbedtls_ssl_get_fatal_alert(const mbedtls_ssl_context *ssl);

```
As it is only for fatal alerts and we only read the SSL context, I propose rather `mbedtls_ssl_get_fatal_alert(const mbedtls_ssl_context *ssl)`. I've also suggested some changes to the documentation.

File: include/mbedtls/ssl.h [hunk 6]
```diff
@@ -1715,6 +1715,13 @@ struct mbedtls_ssl_context {
     int MBEDTLS_PRIVATE(keep_current_message);   /*!< drop or reuse current message
                                                     on next call to record layer? */
 
+    unsigned char MBEDTLS_PRIVATE(in_alert_recv);   /*!< Determines if a fatal alert has
```
Comment (ronald-cron-arm) [inline include/mbedtls/ssl.h]:
```suggestion
    unsigned char MBEDTLS_PRIVATE(in_fatal_alert_recv);   /*!< Determines if a fatal alert has
```

File: include/mbedtls/ssl.h [hunk 7]
```diff
@@ -1715,6 +1715,13 @@ struct mbedtls_ssl_context {
...
+                                                       been received. Values:
+                                                       - \c 0 , no fatal alert received.
+                                                       - \c 1 , a fatal alert has been received */
+    unsigned char MBEDTLS_PRIVATE(in_alert_type); /*!< Type of fatal alert if in_alert_recv
```
Comment (ronald-cron-arm) [inline include/mbedtls/ssl.h]:
```suggestion
    unsigned char MBEDTLS_PRIVATE(in_fatal_alert_type); /*!< Type of fatal alert if in_alert_recv
```

File: library/ssl_msg.c (hunk 2 above)
Comment (ronald-cron-arm) [reply library/ssl_msg.c:5023]: We use MBEDTLS_ERR_SSL_BAD_INPUT_DATA in other APIs to signal that the API is called out of context, like in `mbedtls_ssl_get_own_cid()` or `mbedtls_ssl_get_session()`. Thus using `MBEDTLS_ERR_SSL_BAD_INPUT_DATA` here seems reasonable to me.

**Comment by ng-gsmk** [comment]
Thanks for the review. I [applied your suggestions](https://github.com/Mbed-TLS/mbedtls/pull/10514/commits/15c68993cbdf4eda6bd58191c0fbf56326fb3952) through the github PR interface and fixed the outstanding issues (mostly missing function and property renames) in a [separate commit](https://github.com/Mbed-TLS/mbedtls/pull/10514/commits/4f83ebedd1a19119a4ea1f776092e51eb25675b6):

```
TLS fatal alert getter ............................................ PASS
```

 @minosgalanakis, @ronald-cron-arm: What would be the next steps?

**Comment by ng-gsmk** [comment]
It looks like the  code suggestions contained a [code style violation](https://ci.trustedfirmware.org/job/mbed-tls-pr-head/job/PR-10514-head/15/pipeline-console/?selected-node=568):

```
12:42:28  ******************************************************************
12:42:28  * check_code_style: Check C code style 
12:42:28  * Mon Jan 26 12:42:27 UTC 2026
12:42:28  ******************************************************************
12:42:34  --- include/mbedtls/ssl.h	2026-01-26 12:31:41.000000000 +0000
12:42:34  +++ include/mbedtls/ssl.h.uncrustify	2026-01-26 12:42:33.944903874 +0000
12:42:34  @@ -1716,11 +1716,11 @@
12:42:34                                                       on next call to record layer? */
12:42:34   
12:42:34       unsigned char MBEDTLS_PRIVATE(in_fatal_alert_recv);   /*!< Determines if a fatal alert has
12:42:34  -                                                       been received. Values:
12:42:34  -                                                       - \c 0 , no fatal alert received.
12:42:34  -                                                       - \c 1 , a fatal alert has been received */
12:42:34  +                                                             been received. Values:
12:42:34  +                                                             - \c 0 , no fatal alert received.
12:42:34  +                                                             - \c 1 , a fatal alert has been received */
12:42:34       unsigned char MBEDTLS_PRIVATE(in_fatal_alert_type); /*!< Type of fatal alert if in_alert_recv
12:42:34  -                                                     != 0 */
12:42:34  +                                                           != 0 */
12:42:34   
12:42:34       /* The following three variables indicate if and, if yes,
12:42:34        * what kind of alert is pending to be sent.
12:42:52  include/mbedtls/ssl.h changed - code style is incorrect.
12:42:52  ^^^^check_code_style: Check C code style: ./framework/scripts/code_style.py -> 1^^^^
``` 

So the fix would be to indent it more (till the `<` of the line earlier)? 

```diff
diff --git a/include/mbedtls/ssl.h b/include/mbedtls/ssl.h
index e5f1b64285..247711b15b 100644
--- a/include/mbedtls/ssl.h
+++ b/include/mbedtls/ssl.h
@@ -1716,9 +1716,9 @@ struct mbedtls_ssl_context {
                                                     on next call to record layer? */
 
     unsigned char MBEDTLS_PRIVATE(in_fatal_alert_recv);   /*!< Determines if a fatal alert has
-                                                       been received. Values:
-                                                       - \c 0 , no fatal alert received.
-                                                       - \c 1 , a fatal alert has been received */
+                                                             been received. Values:
+                                                             - \c 0 , no fatal alert received.
+                                                             - \c 1 , a fatal alert has been received */
     unsigned char MBEDTLS_PRIVATE(in_fatal_alert_type); /*!< Type of fatal alert if in_alert_recv
                                                      != 0 */
```

File: ChangeLog.d/alert-getter.txt [hunk 8]
```diff
@@ -0,0 +1,6 @@
+Features
+   * Add the function mbedtls_ssl_get_alert() which returns the
+     last received fatal error alert type for a more generic
+     MBEDTLS_ERR_SSL_FATAL_ALERT_MESSAGE return value from
+     mbedtls_ssl_handshake(), mbedtls_ssl_handshake_step() or
+     mbedtls_ssl_read().
```
Comment (ronald-cron-arm) [inline ChangeLog.d/alert-getter.txt]: The name of the function should be updated. Otherwise what about the following that sounds a bit clearer to me:

Add the function mbedtls_ssl_get_fatal_alert(), which returns the type of
the last received fatal alert. This allows callers to retrieve more
detailed information when mbedtls_ssl_handshake(),
mbedtls_ssl_handshake_step(), or mbedtls_ssl_read() returns the generic
MBEDTLS_ERR_SSL_FATAL_ALERT_MESSAGE error code.
